### PR TITLE
Open multiseries tiff files and nd2 files

### PR DESCRIPTION
**Pull request by RuiOSilva**
# Description

Add functionality to work with multiseries tif files and nd2 (from Nikon softwares) files. 
These changes can allow people that record in Nikon microscopes to quickly use Caiman functions without data conversion. Also, it can allow to open big tiff files that come in multiseries, or to save big files iterativelly for later use using the append variable of the tifffile.TiffFile.imwrite() function, which stores bigTiff data in various series.

These changes requires nd2reader as a dependency (accessible through conda-forge)

The changes are:
1 - Creation of a load_multiseries_tiff_file function in caiman.base.movies. This function allows opening any sets of frames based on slice() defined frames. 
2 - Change of load and load_iter functions in caiman.base.movie() to allow opening these two file types 
3 - Change of get _file_size function in caiman.source_extraction.cnmf.utilities to also account for these two new file types.

# Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# Branching
- All PRs should be made against the **dev** branch. The master branch is not often merged back to dev.
- If you want to get your PR out to the world faster (urgent bugfix), poke pgunn to cut a release; this will get it onto github and into conda faster

# Has your PR been tested?

If you're fixing a bug or introducing a new feature it is recommended you run the tests by typing

```python caimanmanager.py test```

and

```python caimanmanager.py demotest```

prior to submitting your pull request. 

It passed the demotest and I ran the OnAcid notebook with the original supplied data and with data in these two formats without any problems. However it gave the following error while running the python caimanmanager.py test

error_openmultiseries

### CODE REVIEW

File: caiman/source_extraction/cnmf/utilities.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
 import scipy
 from scipy.sparse import spdiags, issparse, csc_matrix, csr_matrix
 import scipy.ndimage.morphology as morph
+from nd2reader import ND2Reader
```
Comment (kushalkolar) [inline caiman/source_extraction/cnmf/utilities.py:36]: you'd want to make this an optional import, users who don't have nd2 files shouldn't have to install it, something like:

```python

try:
    from nd2reader import ND2Reader
except ImportError:
    HAS_ND2 = False
else:
    HAS_ND2 = True
```

File: caiman/source_extraction/cnmf/utilities.py [hunk 2]
```diff
@@ -992,13 +993,23 @@ def get_file_size(file_name, var_name_hdf5='mov'):
...
-                    T, dims = siz[0], siz[1:]
+                    siz = tffl.series[0].shape
+                    # tiff files written in append mode
+                    if len(siz) < 3:
+                        dims = siz
+                        T = len(tffl.pages)
+                    else:
+                        T, dims = siz[0], siz[1:]
+            elif extension == '.nd2':
+                nd2 = ND2Reader(file_name)
```
Comment (kushalkolar) [inline caiman/source_extraction/cnmf/utilities.py:1010]: from above, something like

```python
if not HAS_ND2:
    raise ImportError("nd2reader must be installed to use nd files")
```

**Review by kushalkolar — COMMENTED** [review]
I had some comments, but anyways I'm not the maintainer and I'd wait to see what Pat says

File: caiman/source_extraction/cnmf/utilities.py (hunk 1 above)
Comment (pgunn) [reply caiman/source_extraction/cnmf/utilities.py:36]: It's a pure-python dependency available in conda-forge and pypi; having it as a static dependency is fine (preferable actually)

**Comment by pgunn** [comment]
If we're going down the static dependency route (it's a pretty easy dependency I think?) then we don't want any code that checks to see if it's there and things can be simpler.

Instead, please look into modifying the yaml files and the requirements.txt (and I'll modify the conda-forge plumbing to add it as a dependency to the binary package).

Please note though that if this ever in the future does become a troublesome dependency for any reason we'll be pulling it back out of the code.

**Comment by kushalkolar** [comment]
Do you think it's just safer to keep it as an optional dependency?

@RuiOSilva just curious, were your multi series tiff files openable after this PR: https://github.com/flatironinstitute/CaImAn/pull/992

There's a never ending complexity with tiff files :)

**Comment by pgunn** [comment]
> Do you think it's just safer to keep it as an optional dependency?
> 
> @RuiOSilva just curious, were your multi series tiff files openable after this PR: #992
> 
> There's a never ending complexity with tiff files :)

Optional dependencies are something IMO best avoided unless there's a really good reason; they make code analysis hard and make it unsafe to cleanup things. They also confuse static analysis tools.

**Comment by RuiOSilva** [comment]
@kushalkolar yes, these changes allowed me to run multiseries tiff files even with online CNMF, still not with .mmap though.
I will try to do the PR for the motion correction templates tomorrow. I still have some things to change in my codes.

**Comment by pgunn** [comment]
@RuiOSilva Giving you a heads-up that we're going to be resolving the merge conflict and ideally getting this merged into dev in some form over the next few days. Some parts, like the changes to play(), will be too involved to preserve, but the file format work will go in.

**Comment by pgunn** [comment]
Hello,
Apologies for asking you to do this, but we haven't gotten around to looking at this until recently, and in the meantime the load() and load_iter() functions have needed to be reworked pretty substantially. I've been looking through the code trying to find ways to get your load_iter() changes for tiff in there, but it looks like we changed the base code and translating the intent is pretty difficult.

If you wouldn't mind redoing the diff against dev (maybe break it into two diffs, one for nd2reader, one for the tiff changes), ideally not restructuring the code in general too much (add nd2reader to the bottom of the parse list), I'd love to get these changes in there.

**Pull request closed without merging**